flow/DeterministicRandom.cpp

```diff
@@ -92,9 +92,12 @@ uint64_t DeterministicRandom::randomUInt64() {
 }
 
 uint32_t DeterministicRandom::randomSkewedUInt32(uint32_t min, uint32_t maxPlusOne) {
-	std::uniform_real_distribution<double> distribution(std::log(min), std::log(maxPlusOne - 1));
+	// prevent log(0) error
+	double lhs = min == 0 ? 1e-15 : static_cast<double>(min);
+	double rhs = maxPlusOne <= 1 ? 1e-15 : static_cast<double>(maxPlusOne - 1);
+	std::uniform_real_distribution<double> distribution(std::log(lhs), std::log(rhs));
 	double logpower = distribution(random);
-	uint32_t loguniform = static_cast<uint32_t>(std::pow(10, logpower));
+	uint32_t loguniform = static_cast<uint32_t>(std::pow(M_E, logpower));
```

Comment:
I think the value 0 might be over represented in this distribution. The rough idea of this function is that each order of magnitude is equally represented, and there are a lot of these between 1e-15 and 1. 

I believe it would give good results to generate from:

```
std::uniform_real_distribution<double> distribution(std::log(min+1), std::log(maxPlusOne));
```

and then subtract one from the final result. You’ll also want to assert that min is less than maxPlusOne so that you don’t overflow on the addition (just like the assert in randomInt64).

Reply:
I guess we may also need to address the probability of generating maxPlusOne-1. In the current and my proposed scheme, I think this number may not be possible to get because the end of the log distribution is exclusive. Another alternative is to generate:

```
std::uniform_real_distribution<double> distribution(std::log(std::max<double>(min, 1/M_E)), std::log(maxPlusOne));
```

And then not subtracting 1 at the end. The use of 1/M_E as the lower bound would make 0 equally likely as anything between 1 and M_E, which sounds reasonable, but we could adjust that as needed.

Reply:
Great suggestions! Done.